File: 20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # Create a list as stack
        parentheses_list = []
        for c in s:
            if c in ["(", "{", "["]:
                parentheses_list.append(c)
            else:
                if c == ")":
                    # In the if statement below, it will check the length first, and if it equals 0,
                    # it won't do the .pop() operation because we use "or" here!
                    # It will then directly return False.
                    if len(parentheses_list) == 0 or parentheses_list.pop() != "(":
                        return False
                elif c == "}":
                    if len(parentheses_list) == 0 or parentheses_list.pop() != "{":
                        return False
                elif c == "]":
                    if len(parentheses_list) == 0 or parentheses_list.pop() != "[":
                        return False
        return True if len(parentheses_list) == 0 else False

File: cpp/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        for (const auto& c : s){
            if (c=='(' || c=='{' || c=='['){
                stk.push(c);
            }
            else if (c == ')'){
                if (stk.size() == 0) return false;
                if (stk.top() != '(') return false;
                stk.pop();
            }
            else if (c == '}') {
                if (stk.size() == 0) return false;
                if (stk.top() != '{') return false;
                stk.pop();
            }
            else if (c == ']'){
                if (stk.size() == 0) return false;
                if (stk.top() != '[') return false;
                stk.pop();
            }
        }
        if (stk.size() > 0)
            return false;
        return true;
    }
};